### DISCUSSION TITLE
Use of CTA with DIRAC

### DISCUSSION THREAD

**Discussion by IgorPelevanyuk** [DIRACGrid/DIRAC#7554] [General]
CTA system for tape libraries are getting used more around the world. DIRAC allows integrating CTA storages. But the question is how to get it work efficiently. I know that this topic may be interested for @zhangxiaomei .

So far we tested data upload (to CTA). It works smoothly since when we upload data to CTA, we actually upoload data to a disk buffer, and once the data are there DIRAC consider transfer to CTA successful. CTA will eventually upload data from the buffer to a tape and remove data from the buffer. The method to upload data that we used is just to do 
`dirac-dms-add-file <desired_LFN> <file_path> <CTA_storage_name>`

Main question appears when we want to read that data. For example if we "lost" original data on the disk and want to get them from the tape. 

CTA instruction that we have proposes the following instruction for direct work with CTA:
1. Send request to place data from tape to buffer:
`xrdfs $EOS_MGM_URL prepare -s /eos/project/test/file.txt`

2. Perfom checks until the file is buffered: 
`xrdfs $EOS_MGM_URL query prepare 12 /eos/project/test/file.txt`

3. Once the file is buffered we can download it:
`xrdcp root://cta-eos.jinr-t1.ru//eos/project/user/…. file.txt`
If we lose the moment when the data are placed on buffer, the data may be removed by CTA after some time (configured on CTA side)

So, the following questions apperas:

- Does anybody have already an experience with CTA under DIRAC?
- How to read data from CTA reliably and efficiently?
- Does DIRAC has any service or agent or set of both for this type of operations?
- What is recomended practices for CTA integration in DIRAC? 
- Anything we as administrators and users should know when working with them?

**Comment by zhangxiaomei** [comment]
CERN is using CTA so that I think LHCb and ILC should use them. JUNO has a simple test with CTA on reading.
According to DIRAC documentation (https://dirac.readthedocs.io/en/latest/AdministratorGuide/Resources/storage.html#staging), since CTA only uses root protocol rather than srm protocol and root protocol doesn't support auto-prestage, so the CTA need to prestage data from tape firstly before reading it from disk buffer:
      >>> DM.prestageReplica("/juno/user/j/jpandre/tape_test_202401/tape_test_00200.rnd", "IHEP-TAPE")
    {'OK': True, 'Value': {'Failed': {}, 'Successful': {'/juno/user/j/jpandre/tape_test_202401/tape_test_00200.rnd': 
    '0446ca7a2196:bb417da0.64c70213:1:1704937212'}}}

**Reply by andresailer** [reply]
I think the StorageManagementSystem can stage files on CTA?  Otherwise I use the request management system (in collaboration with the TransformationSystem) to transfer files from CTA to disk storage if they are needed.

**Reply by IgorPelevanyuk** [reply]
@zhangxiaomei 
> >>> DM.prestageReplica("/juno/user/j/jpandre/tape_test_202401/tape_test_00200.rnd", "IHEP-TAPE")

This code looks like python code. I will try to test it and see how it works.
But still, does that mean that there is no DIRAC script for that, like dirac-dms-get-file?

**Reply by IgorPelevanyuk** [reply]
@andresailer 
Does that mean that RequestManagementSystem is aware about staging mechanism and automatically fills CTA with requests for staging, checks the status of all files during staging process, and once any of the requested files appears on the disk RMS initiates a transfer of this particular file?

**Reply by andresailer** [reply]
As long as you do the transfers via FTS, yes.

**Reply by zhangxiaomei** [reply]
We haven't done through FTS yet and also interested to have a try.

**Comment by zhangxiaomei** [comment]
> @zhangxiaomei
> 
> > > > > DM.prestageReplica("/juno/user/j/jpandre/tape_test_202401/tape_test_00200.rnd", "IHEP-TAPE")
> 
> This code looks like python code. I will try to test it and see how it works. But still, does that mean that there is no DIRAC script for that, like dirac-dms-get-file?

We haven't found a proper command to do it. If needed, can write one.

**Comment by IgorPelevanyuk** [comment]
Considering _prestageReplica_ function. I tested it in the following way:
```
import DIRAC.DataManagementSystem.Client.DataManager as DM
dm = DM.DataManager()
dm.prestageReplica("/bmn.nica.jinr/vo/backup/raw/run8/mpd_run_Top_6670_ev1_p7.data", "JINR-CTA-BMN")
{'OK': True, 'Value': {'Failed': {}, 'Successful': {'/bmn.nica.jinr/vo/backup/raw/run8/mpd_run_Top_6670_ev1_p7.data': '04469f5de77e:019ee432.660bf6ae:1:1712298700'}}}
```
This command sent request to place data from tape to buffer. Theoretically, if there would be a command to check if the file appeared on the disk buffer, DIRAC RMS could initiate TPC request to copy data on a storage.

**Comment by IgorPelevanyuk** [comment]
And question about FTS. Does that mean that if some DIRAC installation does not have FTS it will not be possible to work with CTA efficiently?
